x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java

```diff
@@ -31,44 +34,99 @@
 /**
  * Replaces an aggregation function having a false/null filter with an EVAL node.
  * <pre>
- *     ... | STATS x = someAgg(y) WHERE FALSE {BY z} | ...
+ *     ... | STATS/INLINESTATS x = someAgg(y) WHERE FALSE {BY z} | ...
  *     =>
- *     ... | STATS x = someAgg(y) {BY z} > | EVAL x = NULL | KEEP x{, z} | ...
+ *     ... | STATS/INLINESTATS x = someAgg(y) {BY z} > | EVAL x = NULL | KEEP x{, z} | ...
  * </pre>
+ *
+ * This rule is applied to both STATS' {@link Aggregate} and {@link InlineJoin} right-hand side {@link Aggregate} plans.
+ * The logic is common for both, but the handling of the {@link InlineJoin} is slightly different when it comes to pruning
+ * its right-hand side {@link Aggregate}.
  */
-public class ReplaceStatsFilteredAggWithEval extends OptimizerRules.OptimizerRule<Aggregate> {
+public class ReplaceStatsFilteredAggWithEval extends OptimizerRules.OptimizerRule<LogicalPlan> {
     @Override
-    protected LogicalPlan rule(Aggregate aggregate) {
-        int oldAggSize = aggregate.aggregates().size();
-        List<NamedExpression> newAggs = new ArrayList<>(oldAggSize);
-        List<Alias> newEvals = new ArrayList<>(oldAggSize);
-        List<NamedExpression> newProjections = new ArrayList<>(oldAggSize);
+    protected LogicalPlan rule(LogicalPlan plan) {
+        Aggregate aggregate;
+        Holder<InlineJoin> ij = new Holder<>();
+        if (plan instanceof Aggregate a) {
```

Comment:
This code block, in a nutshell, finds out if the `Aggregate` comes from `inlinejoin` or is a standalone one.